pyiron_atomistics/sphinx/parser_base.py

```diff
@@ -0,0 +1,201 @@
+# coding: utf-8
+# Copyright (c) Max-Planck-Institut für Eisenforschung GmbH - Computational Materials Design (CM) Department
+# Distributed under the terms of "New BSD License", see the LICENSE file.
+
+__author__ = "Christoph Freysoldt"
+__copyright__ = (
+    "Copyright 2023, Max-Planck-Institut für Eisenforschung GmbH - "
+    "Computational Materials Design (CM) Department"
+)
+__version__ = "1.0"
+__maintainer__ = "Christoph Freysoldt"
+__email__ = "freysoldt@mpie.de"
+__status__ = "production"
+__date__ = "Dec 8, 2023"
+
+import re
+from types import GeneratorType
+import numpy
+
+
+class KeywordTreeParser:
+    """
+    A base class to parse files block by block via keyword-triggered
+    parsing routines organized in a tree. Parsing routines can
+    add more levels of keyword->parse function maps. The file
+    is read line by line on demand while parsing, so large files will not clobber
+    memory.
+
+    A parser routine can either return or yield (once!) to continue parsing. If it yields,
+    the rest of the routine (after yield) will be executed when the next keyword
+    of the current or a higher level is found.
+
+    Every parser routine MUST remove the keyword from the lineview.
+
+    A typical use will be
+
+    class my_parser(KeywordTreeParser):
+        def __init__(self,file)
+            super ().__init__({
+                    "key1" : self.parse_key1,
+                    "key2" : self.parse_key2  })
+            self.parse (file)
+
+    """
+
+    def __init__(self, keylevels=[]):
+        if isinstance(keylevels, dict):
+            keylevels = [keylevels]
+        elif not isinstance(keylevels, list):
+            raise TypeError
+        self.keylevels = keylevels
+
+    def parse(self, filename):
+        """
+        Parse a file using the current keylevels
+
+        Args:
+            filename ... the filename of the file to parse
+        Returns: nothing
+        """
+        # --- initialization
+        if len(self.keylevels) == 0:
+            raise KeyError("No parsing functions available in keylevels")
+        filehandle = open(filename)
+        # the following properties only exist while parsing
+        self.line = filehandle.__iter__()
+        self.lineview = ""
+        self.filename = filename
+        self.lineno = 0
+        self.line_from = 0
+        while True:
+            for keymap in self.keylevels:
+                for key, func in keymap.items():
+                    if key in self.lineview:
+                        self._cleanup(keymap)
+                        res = func()
+                        if isinstance(res, GeneratorType):
+                            res.send(None)
+                            keymap["%finalize!"] = res
+                        break
```

Comment:
From here on there are wrong indentations with different levels

Reply:
I think it's correct, but is making use of the not very well known `for/else` feature of python.

Reply:
aha ok I didn't know that it was possible